File: Data Structure/Python/linked_list.py
class LinkedList:
    class Node:
        def __init__(self, data = None, next = None):
            self.data = data
            self.next = next
            
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0
        
    def addToTail(self, value):
        self.length += 1
        if self.head == None:
            self.head = self.tail = self.Node(value, None)
        else:
            self.tail.next = self.tail = self.Node(value, None)
            
    def addToHead(self, value):
         self.length += 1
         if self.head == None:
                self.head = self.tail = self.Node(value, None)
         else:
                    self.head = self.Node(value, self.head)
            
    def insertInPos(self, index, value):
        if index < 0 or index > self.length + 1:
            return
        if index == 0:
            addToHead(value)
        if index == self.length + 1:
            addToTail(value)
        curr = self.head
        n = 0
        while curr != None:
            if n == index:
                curr.next = self.Node(value, curr.next)
                break
            ++n
            curr = curr.next
            
    def getLength(self):
        return self.length
    
    def remove(self, value):
        if self.head == None: return
        curr = self.head
        prev = None
        while curr != None:
            if curr.data == value:
                if prev != None:
                    prev.next = curr.next
                    if curr.next == None:
                        tail = prev
                        tail.next =None
                    curr = prev
                else:
                    curr = self.head.next
                    self.head = curr
                    if self.head == None:
                        tail = None
            self.length -= 1
            prev = curr
            curr = curr.next
        
    def search(self, value):
        curr = self.head
        index = 0
        while curr != None:
            if curr.data == value:
                return (True, index)
            curr = curr.next
            index += 1
        return (False, -1)

    def show(self):
        if self.head == None: return
        curr = self.head
        while curr != None:
            print(curr.data, sep=' ')
            curr = curr.next
            
    def __str__(self):
        if self.head == Nonde: return 'LinkedList []'
        curr = self.head
        out = 'LinkedList: '
        while curr != None:
            out += str(curr.data)+' '
            curr = curr.next
        return out  
    
    def sort(self):
        if self.head == None: return
        for i in range(0, self.length - 1):
            tmp = self.head
            while tmp.next != None:
                if tmp.data > tmp.next.data:
                    tmp.data, tmp.next.data = tmp.next.data, tmp.data
                tmp = tmp.next

File: Data Structure/C++/linked_list.cpp
#include <iostream>

template<typename T>
class linkedList
{
    private:
    struct node;
    node* head, *tail;
    size_t count;
    public:
    linkedList();
    ~linkedList();
    void addToTail(const T&);
    void addToHead(const T&);
    void removeFromHead();
    void removeElementByVal(const T&);
    void show();
    bool contain(const T&);
    void insert(node* pos, const T&);
    

};

template<typename T>
struct linkedList<T>::node
{
    T data;
    node* next;
    node() = default;
    node(const T& val) : data(val), next(nullptr){}
};

template<typename T>
linkedList<T>::linkedList() : head(nullptr), tail (nullptr), count(0)
{
}  

template<typename T>
linkedList<T>::~linkedList()
{
    while(head != nullptr)
    {
        node* tmpNewHead = head->next;
        delete head;
        head = tmpNewHead;
    }
    tail = nullptr;
    count = 0;
}

template<typename T>
void linkedList<T>::addToTail(const T& val)
{
    node* newNode = new (std::nothrow) node(val);
    if(newNode != nullptr)
    {
        if(head == nullptr)
        {
            head = tail = newNode;
        } else
        {
            tail -> next = newNode;
            tail = newNode;
            ++count;
        }
    }
}

template<typename T>
void linkedList<T>::addToHead(const T& val)
{
    node* newNode = new (std::nothrow) node(val);
    if(newNode != nullptr)
    {
       newNode -> next = head;
       head = newNode;
       ++count;
    } 
}

template<typename T>
void linkedList<T>::removeFromHead()
{
    if (head != nullptr)
    {
        node* newNode = head -> next;
        delete head;
        head = newNode;
    }
}

template<typename T>
void linkedList<T>::removeElementByVal(const T& val)
{
    if (head == nullptr) return;
    
    node *current = head, *prev = nullptr;
    while( current != nullptr)
    {
        if(current -> data == val)
        {
            if(prev != nullptr)
            {
               prev -> next = current -> next;
                if(current -> next == nullptr)
                {
                    tail = prev;
                    tail -> next == nullptr;
                }
                delete current;
                current = prev;
            }
            else
            {
                current = head -> next;
                delete head;
                head = current;
                if(head == nullptr)
                {
                    tail = nullptr;
                }
                
            }
            --count;
        }
        prev = current;
        current = current -> next;
    }
}

template<typename T>
void linkedList<T>::show()
{
    node *tmp = head;
    while(tmp != nullptr)
    {
        std::cout<<tmp -> data<<" ";
        tmp = tmp -> next;
    }
    std::cout<<std::endl;
}

template<typename T>
bool linkedList<T>::contain(const T &val)
{
    if(head == nullptr) return false;
    node* tmp = head;
    while(tmp != nullptr)
    {
        if( tmp -> data == val) return true;
        tmp = tmp -> next;
    }
    return false;
}
template<typename T>
void linkedList<T>::insert(node *pos, const T &val)
{
    if(pos == nullptr) return;
    node *curr = head, *prev = nullptr;
    while( curr != pos)
    {
        prev = curr;
        curr = curr -> next;
    }
    
    node* newNode = new (std::nothrow) node(val);
    if(newNode != nullptr)
    {
        if(pos == head)
        {
            newNode -> next = head;
            head = newNode;
        } else
        {
            newNode -> next = pos;
            prev -> next = newNode;
        }
    }
    ++count;
}
    
int main()
{
 
}

